include/aws/s3/private/s3_checksums.h

```diff
@@ -0,0 +1,102 @@
+#ifndef AWS_S3_Checksums_H
+#define AWS_S3_Checksums_H
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "aws/s3/s3_client.h"
+#include <aws/cal/hash.h>
+#include <aws/io/stream.h>
+
+#define AWS_CRC32_LEN 4
+#define AWS_CRC32C_LEN 4
+
+struct aws_checksum;
+
+struct aws_checksum_vtable {
+    void (*destroy)(struct aws_checksum *checksum);
+    int (*update)(struct aws_checksum *checksum, const struct aws_byte_cursor *buf);
+    int (*finalize)(struct aws_checksum *checksum, struct aws_byte_buf *out, size_t truncate_to);
+};
+
+struct aws_checksum {
+    struct aws_allocator *allocator;
+    struct aws_checksum_vtable *vtable;
+    void *impl;
+    size_t digest_size;
+    bool good;
```

Comment:
Why is this necessary?

After looking further, I'd rename this based on its intent (calculation complete).